protocol/0083-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,337 @@
+  - the teams **referees** and their current number of epochs in the team
+- a list of all **teams** (by `id`) and the following metrics:
+  - current `team_running_volume` (value at the start of the epoch)
+  - current `referral_reward_factor` applied to referee taker fees
+  - current `referral_discount_factor` applied to referee taker fees
+  - the total amount rewarded to the referrer during the referral program
+  - the total amount discounted for all referees during the referral program
+  - any data required for additional reward metrics
+
+The Trades API should now also expose the following additional information for every trade:
+
+- Referral program rewards
+  - `infrastructure_fee_referral_reward`
+  - `liquidity_fee_referral_reward`
+  - `maker_fee_referral_reward`
+- Referral program discounts
+  - `infrastructure_fee_referral_discount`
+  - `liquidity_fee_referral_discount`
+  - `maker_fee_referral_discount`
+
+## Acceptance Criteria
+
+### Governance Proposals
+
+1. If an `UpdateReferralProgram` proposal does not fulfil one or more of the following conditions, the proposal should be `STATUS_REJECTED`:
+    - the `closing_timestamp` must be less than or equal to the proposals `enactment_time`.
+    - the number of tiers in `benefit_tiers` must be less than or equal to the network parameter `referralProgram.maxBenefitTiers`.
+    - all `minimum_epochs_in_team` values must be an integer strictly greater than 0.
+    - all `referral_reward_factor` values must be greater than or equal to `0` and less than or equal to the network parameter `referralProgram.maxReferralRewardFactor`.
+    - all `referral_discount_factor` values must be greater than or equal to `0` and be less than or equal to the network parameter `referralProgram.maxReferralDiscountFactor`.
+    - the `window_length` must be an integer strictly greater than zero.
+1. A referral program should be started the first epoch change after the `enactment_datetime` is reached.
+1. A referral program should be closed the first epoch change after the `closing_timestamp` is reached.
+1. If a referral program is already active and a proposal `enactment_datetime` is reached, the referral program is updated at the next epoch change.
+    - Propose program A with `enactment_timestamp` 1st Jan and `closing_timestamp` 31st Dec.
+    - Proposal for program A accepted and begins first epoch after 1st Jan.
+    - Propose program B with `enactment_timestamp` 1st June and `closing_timestamp` 31st Aug.
+    - Proposal for program B accepted and overrides program A the first epoch after 1st June.
+    - Program is closed first epoch after 31st Aug, there should be no active proposals.
+
+### Team Mechanics
+
+#### Creating / updating teams
+
+1. If a party **is not** currently a referrer, the party can **create** a team, by submitting a signed `CreateTeam` transaction.
+1. If a party **is** currently a referrer, the party can **update** a team, by submitting a signed `CreateTeam` transaction.
```

Comment:
I would rather have a dedicated payload for that `UpdateTeams` that can have the `teamID` a property (which also mean the team ID / referral code is immutable). Is there anything preventing that ? Is it possible to change the referral code ?

Reply:
If that is preferable then I can update the spec with that separate `UpdateTeam` transaction.

The `teamID` / referral code for a team should never change.